File: Grind75/60-LetterCombinationsPhone/python/LetterCombinationsPhone.py
#!/bin/python3

from __future__ import annotations
import sys
from typing import Dict
sys.path.append('../../Utility/python')
from PrettyPrint import *
from Colors import *

class Solution:
  def letterCombinations(self, digits: str) -> List[str]:
    if digits == '':
      return []
    digitLetters: Dict[str, str] = {
                    '2':'abc', '3':'def', 
        '4':'ghi',  '5':'jkl', '6':'mno',
        '7':'pqrs', '8':'tuv', '9':'wxyz',
    }
    combinations: List[str] = ['']
    for digit in digits:
      letters = digitLetters[digit]
      newCombs: List[str] = []
      for comb in combinations:
        for letter in letters:
          newCombs.append(comb + letter)
      combinations = newCombs
    return combinations


if __name__ == '__main__':
  soln = Solution()
  testCases = [
    ( '', [] ),
    ( '2', ['a','b','c',] ),
    ( '27', ['ap','bp','cp','aq','bq','cq','ar','br','cr','as','bs','cs',] ),
  ]

  for digits, answer in testCases:
    colorPrint(LIGHT_BLUE, digits)
    colorPrint(WHITE, answer)
    result = soln.letterCombinations(digits)
    colorPrint(GREEN if set(result) == set(answer) else RED, result)

File: Grind75/60-LetterCombinationsPhone/cpp/LetterCombinationsPhone.cpp
#include "LetterCombinationsPhone.h"
#include <vector>
#include <string>

using namespace std;

vector<string> Solution::letterCombinations(string digits) {
  if (digits.empty())
    return {};
  vector<string> combinations {""};
  for (const char digit : digits) {
    const string &letters = num2let[digit - '0'];
    vector<string> newCombs;
    for (string &comb : combinations)
      for (const char letter : letters)
        newCombs.push_back(comb + letter);
    combinations = std::move(newCombs);
  }
  return combinations;
}

